### PR TITLE
feat: Add instance templates support (create/get/exists)

### LINKED ISSUE #387: Instance Templates
Instance Templates are still missing from this library.

### PR DESCRIPTION
**Pull request by hermanbanken**
Adds basic Instance Template support: get/create/exists.

----

- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/nodejs-compute/issues/387)
- [x] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #387 🦕

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -360,6 +361,72 @@ class Compute extends common.Service {
       }
     );
   }
+  /**
+   * Create an instance template.
+   *
+   * @see [Instance Templates: insert API Documentation]{@link https://cloud.google.com/compute/docs/reference/v1/instanceTemplates/insert}
+   *
+   * @param {string} name - The name of the target image.
+   * @param {object} parameters - The settings of the template, see the
```
Comment (stephenplusplus) [inline src/index.js]: I don't believe this is in the code. It looks like it's just "name" and "options".

File: src/index.js [hunk 2]
```diff
@@ -360,6 +361,72 @@ class Compute extends common.Service {
...
+   * @param {?error} callback.err - An error returned while making this request.
+   * @param {Operation} callback.operation - An operation object that can be used
+   *     to check the status of the request.
+   * @param {object} callback.apiResponse - The full API response.
+   *
+   * @example
+   * const Compute = require('@google-cloud/compute');
+   * const compute = new Compute();
+   *
+   * const options = {
```
Comment (stephenplusplus) [inline src/index.js]: Just calling this out so something can be thought of to put here :)

File: src/index.js [hunk 3]
```diff
@@ -2156,6 +2223,20 @@ class Compute extends common.Service {
...
+   * @see [Instance Templates Overview]{@link https://cloud.google.com/compute/docs/instanceTemplates}
+   *
+   * @param {string} name - Name of the instance template.
+   * @returns {Image}
+   *
+   * @example
+   * var instanceTemplate = gce.instanceTemplate('instance-template-name');
+   */
+  instanceTemplate(name) {
+    return new InstanceTemplate(this, name);
```
Comment (stephenplusplus) [inline src/index.js]: Could you add a `throw(new Error('An instance template name is required.'))` here if `name` isn't provided?

File: src/index.js [hunk 4]
```diff
@@ -360,6 +361,72 @@ class Compute extends common.Service {
...
+   *
+   * //-
+   * // If the callback is omitted, we'll return a Promise.
+   * //-
+   * compute.createInstanceTemplate('new-image', options).then(function(data) {
+   *   var operation = data[0];
+   *   var apiResponse = data[1];
+   * });
+   */
+  createInstanceTemplate(name, options, callback) {
```
Comment (stephenplusplus) [inline src/index.js]: This method needs tests.

File: src/index.js [hunk 5]
```diff
@@ -2156,6 +2223,20 @@ class Compute extends common.Service {
...
+   *
+   * @see [Instance Templates Overview]{@link https://cloud.google.com/compute/docs/instanceTemplates}
+   *
+   * @param {string} name - Name of the instance template.
+   * @returns {Image}
+   *
+   * @example
+   * var instanceTemplate = gce.instanceTemplate('instance-template-name');
+   */
+  instanceTemplate(name) {
```
Comment (stephenplusplus) [inline src/index.js]: This method needs tests.

**Review by stephenplusplus — CHANGES_REQUESTED** [review]

**Comment by stephenplusplus** [comment]
@hermanbanken Thank you very much for helping out! Everything looks great, I just had a few comments. Feel free to get to them when you have the chance, otherwise we can step in from our side and help as well!

**Comment by bcoe** [comment]
@hermanbanken If you'd be up for addressing @stephenplusplus' review, I would happily land this :smile:

**Comment by JustBeYou** [comment]
@hermanbanken are you still working on this? I'd really like this feature to be in the library and maybe I could help if you don't have the time to finish it

**Comment by JustBeYou** [comment]
@bcoe Could I address the review in a new PR? As I see the original developer is missing.

**Comment by bcoe** [comment]
@JustBeYou 👋 I'll sync up with @stephenplusplus on this issue, I would like to make sure that @hermanbanken gets the commit for the work they've done, perhaps you could submit a patch that against their branch?

**Comment by hermanbanken** [comment]
👋 sorry, I missed the replies

File: src/index.js (hunk 1 above)
Comment (hermanbanken) [reply src/index.js]: Fixed!

File: src/index.js (hunk 3 above)
Comment (hermanbanken) [reply src/index.js]: Fixed

File: src/index.js (hunk 5 above)
Comment (hermanbanken) [reply src/index.js]: Does it? I'm mostly interested in testing `new InstanceTemplate` which is tested in test/instance-template.js

File: src/index.js (hunk 4 above)
Comment (hermanbanken) [reply src/index.js]: Does it? I'm mostly interested in testing new InstanceTemplate which is tested in test/instance-template.js

File: src/index.js (hunk 5 above)
Comment (stephenplusplus) [reply src/index.js]: Yes, we test every method, so we would need a new describe block in test/index.ts that checks:

- should throw if an instance template name is not provided
  - catches the error and validates the contents of the message are as written in the source
- should properly create and return an InstanceTemplate issue
  - confirms that the arguments are correct
  - mocks "InstanceTemplate" and confirms what is returned in the mock is returned to the caller

File: src/index.js (hunk 4 above)
Comment (stephenplusplus) [reply src/index.js]: Yes, we would want something like:

- should make the correct request
  - stubs "this.request"
  - confirms the method is "POST"
  - confirms the URI is "/global/instanceTemplates"
  - confirms the JSON body is `{name: name}`
- should send options in request body
  - stubs "this.request" and confirms the user-provided options are part of the request body
- errors
  - should return the error and API response
    - mocks "this.request" and returns an error and API response
    - confirms the error and API response argument returned are what the caller received
- success
  - should return an instance template
    - mocks "this.instanceTemplate", confirming the "name" received is what the user provided
    - mocks "this.operation", confirming an operation was created from the API response object's "name" property
    - confirms `metadata` was assigned with the API response
    - confirms the user receives the response arguments in order correctly to the callback

**Review by JustinBeckwith — APPROVED** [review]

**Comment by JustinBeckwith** [comment]
@hermanbanken thanks for the PR!  Would love to land this, but the following docs url is 404ing for us:
https://cloud.google.com/compute/docs/instanceTemplates

Could I trouble you to update it with an appropriate link?

**Comment by hermanbanken** [comment]
I don’t have access to my computer ATM, but this is the link: https://cloud.google.com/compute/docs/reference/rest/v1/instanceTemplates.

I adapted a link of another kind, so maybe that link is also broken..

**Comment by bcoe** [comment]
@hermanbanken  anything I can do to help dig into this, wondering where we landed with this.

**Comment by hermanbanken** [comment]
@JustinBeckwith is this no longer wanted? Everyone was looking at me, but I was looking for guidance how to get this merged (I can't merge/approve it myself!).

**Comment by JustinBeckwith** [comment]
Apologies!  I was doing some cleanup, and it had looked like this was abandoned (my mistake).  @stephenplusplus could I trouble you to give another review here?

File: src/index.js [hunk 6]
```diff
@@ -370,6 +371,69 @@ class Compute extends common.Service {
       }
     );
   }
+  /**
+   * Create an instance template.
+   *
+   * @see [Instance Templates: insert API Documentation]{@link https://cloud.google.com/compute/docs/reference/v1/instanceTemplates/insert}
```
Comment (hermanbanken) [inline src/index.js:377]:
```suggestion
   * @see [Instance Templates: insert API Documentation]{@link https://cloud.google.com/compute/docs/reference/rest/v1/instanceTemplates/insert}
```

File: src/index.js [hunk 7]
```diff
@@ -370,6 +371,69 @@ class Compute extends common.Service {
...
+   *
+   * @param {string} name - The name of the target image.
+   * @param {object} [options] - See the
+   *     [Instance Templates: insert API documentation](https://cloud.google.com/compute/docs/reference/v1/instanceTemplates/insert).
```
Comment (hermanbanken) [inline src/index.js:381]:
```suggestion
   *     [Instance Templates: insert API documentation](https://cloud.google.com/compute/docs/reference/rest/v1/instanceTemplates/insert).
```

File: src/instance-template.js [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+    /**
+     * @name InstanceTemplate#formattedName
+     * @type {string}
+     */
+    this.formattedName = InstanceTemplate.formatName_(compute, name);
+  }
+  /**
+   * Delete the instance template.
+   *
+   * @see [InstanceTemplates: delete API Documentation]{@link https://cloud.google.com/compute/docs/reference/v1/instanceTemplates/delete}
```
Comment (hermanbanken) [inline src/instance-template.js:159]:
```suggestion
   * @see [InstanceTemplates: delete API Documentation]{@link https://cloud.google.com/compute/docs/reference/rest/v1/instanceTemplates/delete}
```

**Comment by stephenplusplus** [comment]
@hermanbanken I had some requests for tests here:

- https://github.com/googleapis/nodejs-compute/pull/388#discussion_r399414213
- https://github.com/googleapis/nodejs-compute/pull/388#discussion_r399410276

Are you willing to take that on?

**Comment by hermanbanken** [comment]
I'm not used to this test structure and have the feeling that these tests would be a near duplicate of other types, like
https://github.com/googleapis/nodejs-compute/blob/master/test/instance-group.js
Is that how to go about this? Copy existing test & just adapt?

Given the lack of interest so far, I'm guessing that there is little overlap between the people using this library and people using instanceTemplates.

I'm actually quite busy and don't have the time to pick this up myself. If you - as a maintainer - feel like this is not a valuable addition to the library, please voice that and indeed close the PR. If instead we strive for completeness of this library, anyone may continue.

**Comment by stephenplusplus** [comment]
Yes, copying and pasting the tests should be most of the work. Getting those access paths tested is required to complete your very appreciated work, so if anyone else is interested in picking that up (@JustBeYou ?), that would be appreciated as well. A new PR would be fine.

Thank you again for your help and contributions of code and time to our library. If nobody picks this up soon, I will be happy to finish it up.

**Pull request closed without merging**